bookstore/tests/test_publish.py

```diff
@@ -1,9 +1,109 @@
 import asyncio
+import json
+
+from unittest.mock import Mock
+
 import pytest
 
-from bookstore.publish import BookstorePublishHandler
+from bookstore.publish import BookstorePublishAPIHandler
+from nbformat.v4 import new_notebook
+from tornado.testing import AsyncTestCase, gen_test
+from tornado.web import Application, HTTPError
+from tornado.httpserver import HTTPRequest
+from traitlets.config import Config
 
 
 def test_create_publish_handler_no_params():
     with pytest.raises(TypeError):
-        assert BookstorePublishHandler()
+        assert BookstorePublishAPIHandler()
+
+
+class TestPublishAPIHandler(AsyncTestCase):
```

Comment:
Perhaps rename class to: `MockPublishAPIHandler(AsyncTestCase)`

Reply:
I'm now seeing that the tests are done within the class. @mpacer @MSeal, is the typical pattern to do it this way with S3? I would have expected creating a MockPublishAPIHandler and then running the tests outside of the Mock handler.

Reply:
This is how we're avoiding the whole s3 mocking while still unit testing our handlers. This comes from the model given by tornado. In these cases we're needing to build the handler slightly differently each time because of how it is instantiated on the Tornado side.

Reply:
figuring out this model & how to work with it honestly has been the hardest conceptual part of the entire unit-testing project, as we hadn't found anyone else that was doing true unit tests on their handlers (i.e., without making a mock server, without handling routing through a real application, without sending any actual http requests, &c.).

Reply:
Makes sense. We create a Mock server on JupyterHub even though we are using Tornado. If you and Matt are in agreement on the approach, I'm happy with it. It was more curiosity on my part.

Reply:
i'd tried to follow through the JupyterHub code base & got a lot of really great pointers on how to approach this, but ended up not being able to use too many of the patterns since it did run a a server to handle the requests.

I'm definitely open to taking that approach in the future, if you'd like — it would probably simply a lot of the weirder workarounds we've been doing at the sacrifice of making our unit tests a little more integration-y.

If we wanted to do that, I would really appreciate your guidance around the JupyterHub testing base, because while I was able to learn a lot… its really complicated & was tricky to see how we'd adapt it to our use cases.

Reply:
I'm good with this approach. There's still a few patterns like this where it's awkward / difficult to get in a pure pytest pattern. We could probably reverse engineer a complicated fixture / context wrapper pattern to emulate all the pieces, but given the work required to get the tests running at all in this repo I don't think it'd be worth the effort to go through that additional pain.